### PR TITLE
Cache serverinfo responses

### PR DESCRIPTION
**Pull request by Learath2**
I'm not sure I like it but it works

### CODE REVIEW

File: src/engine/server/server.cpp [hunk 1]
```diff
@@ -1502,8 +1504,63 @@ void CServer::SendServerInfoConnless(const NETADDR *pAddr, int Token, int Type)
...
+}
+
+CServer::CCache::~CCache()
+{
+	Clear();
+}
+
+void CServer::CCache::AddChunk(const void *pData, int Size)
+{
+	SCacheChunk *pNew = new SCacheChunk;
```
[the remark this answers is not in the document]
Comment (Learath2) [reply src/engine/server/server.cpp]: Old C and teeworlds habits. I'll migrate to a `std::list`

File: src/engine/server/server.cpp [hunk 2]
```diff
@@ -1502,8 +1504,63 @@ void CServer::SendServerInfoConnless(const NETADDR *pAddr, int Token, int Type)
...
+		pChunk = pChunk->m_pNext;
+		free(pTmp->m_pData);
+		delete pTmp;
+	}
+
+	m_pRoot = 0;
+	m_pTail = 0;
+}
+
+void CServer::CacheServerInfo(CCache *pCache, int Type, bool SendClients)
```
[the remark this answers is not in the document]
Comment (Learath2) [reply src/engine/server/server.cpp]: There was already an `UpdateServerInfo` which I modified to call `CacheServerInfo` for all caches.
I could rename this to `UpdateServerInfo` but then I'd need to modify it to loop over all caches if `!pCache`.

Not sure which is uglier.

File: src/engine/server/server.cpp [hunk 3]
```diff
@@ -1688,36 +1728,103 @@ void CServer::SendServerInfo(const NETADDR *pAddr, int Token, int Type, bool Sen
...
+	#define ADD_RAW(p, x) (p).AddRaw(x, sizeof(x))
+	#define ADD_INT(p, x) do { str_format(aBuf, sizeof(aBuf), "%d", x); (p).AddString(aBuf, 0); } while(0)
+
+	switch(Type)
 	{
-		if(m_aClients[i].m_State != CClient::STATE_EMPTY)
+	case SERVERINFO_EXTENDED: ADD_RAW(p, SERVERBROWSE_INFO_EXTENDED); break;
+	case SERVERINFO_64_LEGACY: ADD_RAW(p, SERVERBROWSE_INFO_64_LEGACY); break;
+	case SERVERINFO_VANILLA: ADD_RAW(p, SERVERBROWSE_INFO); break;
+	case SERVERINFO_INGAME: ADD_RAW(p, SERVERBROWSE_INFO); break;
```
[the remark this answers is not in the document]
Comment (Learath2) [reply src/engine/server/server.cpp]: Yep

**Comment by Learath2** [comment]
Also, I'm pretty unfamiliar with the performance of STL so I'm not sure if I followed the best practices here. Feel free to let me know

File: src/engine/server/server.cpp [hunk 4]
```diff
@@ -1502,8 +1504,42 @@ void CServer::SendServerInfoConnless(const NETADDR *pAddr, int Token, int Type)
...
+}
+
+CServer::CCache::~CCache()
 {
+	Clear();
+}
+
+CServer::CCache::CCacheChunk::CCacheChunk(const void *pData, int Size)
+{
+	m_pData = malloc(Size); // This can fail, but no way to indicate that barring an exception
```
Comment (def-) [inline src/engine/server/server.cpp]: How about unsigned char m_aData[NET_MAX_PACKETSIZE] or something like that instead of malloc?

Comment (Learath2) [reply src/engine/server/server.cpp]: Ah great idea

File: src/engine/server/server.h [hunk 5]
```diff
@@ -299,10 +299,9 @@ class CServer : public IServer
 		public:
 			CCacheChunk(const void *pData, int Size);
 			CCacheChunk(const CCacheChunk &) = delete;
-			~CCacheChunk();
 
 			int m_DataSize;
-			void *m_pData;
+			unsigned char *m_aData[NET_MAX_PAYLOAD];
```
Comment (def-) [inline src/engine/server/server.h]: Now you're storing pointers, so m_aData is much too big. lose the `*`.

**Pull request merged**